### ISSUE TITLE
[ARM] Incorrect parsing of NEON instructions targeting individual lanes

### ISSUE THREAD

**Issue by dgazzoni** [RRZE-HPC/OSACA#89]
See title. For example, this is the result of parsing `mov.d x21, v23[0]`:

```
{
    "instruction_form": {
        "instruction": "mov.d",
        "operands": [
            {"register": {"prefix": "x", "name": "21"}},
            {"register": {"prefix": "v", "name": "23"}},
            {
                "memory": {
                    "offset": {"value": 0},
                    "base": None,
                    "index": None,
                    "scale": 1,
                }
            },
        ],
        "directive": None,
        "comment": None,
        "label": None,
        "line": "               \tmov.d\tx21, v23[0]",
        "line_number": 19,
        "semantic_operands": {
            "source": [
                {"register": {"prefix": "v", "name": "23"}},
                {
                    "memory": {
                        "offset": {"value": 0},
                        "base": None,
                        "index": None,
                        "scale": 1,
                    }
                },
            ],
            "destination": [{"register": {"prefix": "x", "name": "21"}}],
            "src_dst": [],
        },
        "flags": ["performs_load", "lt_unknown", "tp_unknown"],
        "port_pressure": [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0],
        "port_uops": [],
        "throughput": 0.0,
        "latency": 0.0,
        "latency_wo_load": 0.0,
        "latency_cp": 0,
        "latency_lcd": 0,
    }
}
```

As can be seen, `[0]` is treated as a memory operand rather than its actual purpose of selecting one of the lanes from the NEON register.

This should also affect other instructions, such as `DUP`, multiply/multiply-accumulate/multiply-subtract instructions (floating-point, integer, saturating), `INS`, `LD[1234]`, `MOV`, `SDOT`, `ST[1234]`, `SUDOT`, `USDOT` and possibly others I've missed.

Perhaps, as a simplification, OSACA could ignore dependencies on individual lanes and just declare this to be a dependency on the full NEON register.

**Comment by dgazzoni** [comment]
After going through OSACA's parsing code, I noticed that it does parse lane selectors, but expects them in a different format (I'm feeding the parser from the output of `objdump` in macOS), i.e. something like `mov x21, v23.d[0]`. I've solved this for now in my code using a regex to rewrite the syntax of the affected instructions.

**Comment by JanLJL** [comment]
Hi @dgazzoni,

this is indeed something done in the parser, but I am currently hesitating to implement a fix for this because I cannot reproduce the behavior (unfortunately, I don't have a Mac available).
If I try to compile code with your instruction form, all compilers I can use throw an error stating an `unknown mnemonic` and `unexpected characters following instruction at operand 2`.
Can you provide me with more information about the compiler and OS you used? Also, which flags are you using for objdump?

Thanks!

**Comment by dgazzoni** [comment]
This appears to be a quirk of Apple's syntax for aarch64 assembly, and probably not the only one. Here's the requested info:

```
% sw_vers
ProductName:		macOS
ProductVersion:		13.2
BuildVersion:		22D49

% clang --version
Apple clang version 14.0.0 (clang-1400.0.29.202)
Target: arm64-apple-darwin22.3.0
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin

% objdump --version
Apple LLVM version 14.0.0 (clang-1400.0.29.202)
  Optimized build.
  Default target: arm64-apple-darwin22.3.0
  Host CPU: apple-a12

  Registered Targets:
    aarch64    - AArch64 (little endian)
    aarch64_32 - AArch64 (little endian ILP32)
    aarch64_be - AArch64 (big endian)
    arm        - ARM
    arm64      - ARM64 (little endian)
    arm64_32   - ARM64 (little endian ILP32)
    armeb      - ARM (big endian)
    thumb      - Thumb
    thumbeb    - Thumb (big endian)
    x86        - 32-bit X86: Pentium-Pro and above
    x86-64     - 64-bit X86: EM64T and AMD64

% echo "mov.d x21, v23[0]\numov x21, v23.d[0]" > test.S

% clang -c test.S

% objdump -d test.o

test.o:	file format mach-o arm64

Disassembly of section __TEXT,__text:

0000000000000000 <ltmp0>:
       0: f5 3e 08 4e  	mov.d	x21, v23[0]
       4: f5 3e 08 4e  	mov.d	x21, v23[0]
```

**Comment by JanLJL** [comment]
Ok, with clang I could verify the correctness as well, thank you!

I now implemented the parser to take indexed registers without a shape/lanes and applied the database lookup in a similar way to the GAS suffixes. Whenever there is a mnemonic with a "`.`", i.e., this also includes condition codes such as `eq`, `ne`, etc... , OSACA checks first for the specific instruction and - if nothing is found - then for the instruction excluding the suffix, i.e., `b.ne` falls back to `b`, `mov.d` and `mov.s` fall back to `mov` if not specified explicitly in the DB.
All of this is included from commit 9f715c0 onwards.

If this isn't sufficient in your eyes, please let me know and we can discuss other alternatives!
